### PR TITLE
[MoE] Enable Shared/Routed Overlap For Latent MoE (Nemotron-H)

### PR DESCRIPTION
**Pull request by danielafrimi**
Enable parallel CUDA stream execution between shared and routed experts for latent MoE architectures (e.g., Nemotron-H).

## Problem

Latent MoE compresses the input before routing to experts:
- **Routed experts** receive compressed input: `[S, moe_latent_size]` (e.g., 1024)
- **Shared experts** need original input: `[S, hidden_size]` (e.g., 4096)

Previously, this dimension mismatch prevented the streaming optimization where shared and routed experts run in parallel on separate CUDA streams.

**Before (Sequential - No Streaming in nemotron_h.py):**
```
Time ────────────────────────────────────────────────────────────────►

GPU:  [fc1_latent_proj] ──► [routed_experts] ──► [shared_experts] ──► [combine]
      [S,4096]->[S,1024]     [S,1024]             [S,4096]
      
      |<──────────────────── Total Time ────────────────────────────>|
```

## Solution

Save the original `hidden_states` before applying the transform, allowing shared experts to use the original while routed experts use the transformed input:

**After (Parallel Streaming):**
```
Time ────────────────────────────────────────────────────────────────►

MAIN     ┌──────────────────────────────────────────┐
STREAM   │ [fc1_latent_proj] ──► [routed_experts]   │──► [combine]
         │ [S,4096]->[S,1024]     [S,1024]          │
         │                                          │
         │            (in parallel)                 │
         │                                          │
AUX      │ [shared_experts]                         │
STREAM   │ [S,4096] -> [S,4096]                     │
         └──────────────────────────────────────────┘
                                                   SYNC
         
         |<──────────── Reduced Time ─────────────>|
```

## Implementation Flow (Implemented it on NemotronH for example)

```
NemotronHMoE.forward(hidden_states=[S, 4096])
    |
    v
SharedFusedMoE(routed_input_transform=fc1_latent_proj)
    |
    v
FusedMoE.forward_native()
    |
    |-- Save: _shared_experts_input = hidden_states     [S, 4096]
    |-- Transform: hidden_states = fc1_latent_proj()    [S, 1024]
    |
    v
FusedMoE.forward_impl()
    |
    |-- Clone _shared_experts_input for AUX stream
    |
    +---------------------------+---------------------------+
    |       MAIN STREAM         |        AUX STREAM         |
    |                           |                           |
    |   router(hidden_states)   |                           |
    |   routed_experts          |   shared_experts          |
    |   [S, 1024]               |   [S, 4096]               |
    +---------------------------+---------------------------+
                                |
                               SYNC
                                |
                                v
    Return (shared_output [S, 4096], routed_output [S, 1024])
```

## Changes

| File | Change |
|------|--------|
| `layer.py` | Add `apply_routed_input_transform()` hook in `FusedMoE` base class |
| `shared_fused_moe.py` | Add `routed_input_transform` parameter, override hook |
| `nemotron_h.py` | Pass `routed_input_transform=fc1_latent_proj` to `SharedFusedMoE` |

## Benchmark Results Summary (H100, TP=4)

  | Test | Batch | ISL | OSL | Streaming ON (tok/s) | Streaming OFF (tok/s) | Difference | Verdict |
  |------|-------|-----|-----|----------------------|------------------------|------------|---------|
  | 1 | 1 | 100 | 2000 | 64.71 | 65.04 | -0.5% | Neutral |
  | 2 | 4 | 100 | 1000 | 258.61 | 250.14 | **+3.4%** | ✅ Helps |
  | 3 | 128 | 100 | 2000 | 3701.34 | 3701.12 | ~0% | Neutral |
  | 4 | 128 | 500 | 4000 | 3446.49 | 3443.06 | **+0.1%** | Neutral |

  ## Serving Benchmark Results (B200, 8K Input / 64K Output)

  | Metric | Streaming ON | Streaming OFF | Difference |
  |--------|--------------|---------------|------------|
  | Output token throughput (tok/s) | 4989.99 | 4986.64 | +0.07% |
  | Mean TTFT (ms) | 27157.10 | 26324.07 | +3.2% |
  | Mean TPOT (ms) | 25.22 | 25.25 | -0.1% |

  **Verdict:** Neutral (~0% throughput difference)

  **Config:** 256 requests, 128 max concurrency, 8K ISL / 64K OSL

### CODE REVIEW

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 1]
```diff
@@ -354,6 +354,10 @@ def __init__(
                     "Enabled separate cuda stream for MoE shared_experts", scope="local"
                 )
 
+        # For latent MoE: stores original hidden_states before routed_input_transform
+        # so shared_experts can use it for cloning (they need original dimension)
+        self._shared_experts_input = None
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/layer.py]: Could you consider using a context manager to manage this state instead of doing it with an instance variable? I think it would be more robust, since currently it isn't cleared in error paths and would not be thread safe.

Alternately you could plumb it down as an argument down the call chain but I think that would be too invasive.

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 2]
```diff
@@ -2033,19 +2082,34 @@ def moe_forward_shared(
...
     layer_name: str,
+    shared_experts_input: torch.Tensor | None = None,
 ) -> tuple[torch.Tensor, torch.Tensor]:
     self = get_layer_from_name(layer_name)
     assert self.shared_experts is not None
+
+    # Set here because torch.compile skips forward_native() setup code
+    # and calls this op directly. forward_impl() reads from this var.
+    self._shared_experts_input = shared_experts_input
     return self.forward_impl(hidden_states, router_logits)
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/layer.py]: should `_shared_experts_input` be cleared here prior to returning?

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 3]
```diff
@@ -1549,6 +1571,15 @@ def forward_native(
...
+        # For latent MoE: save ORIGINAL hidden_states before transform
+        # (shared_experts need original dimension, routed experts use transformed)
+        original_hidden_states = hidden_states
+        og_shared_hidden_states = hidden_states.shape[-1]
+
+        # Apply transform for routed experts (e.g., latent projection for latent MoE)
+        hidden_states = self.apply_routed_input_transform(hidden_states)
+
+        # This is the dimension after transform (for routed expert output slicing)
         og_hidden_states = hidden_states.shape[-1]
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/layer.py]: Could we do a bit of renaming here? Having `og_shared_hidden_states` and `og_hidden_states` both represent dimensions while `original_hidden_states` is a tensor is confusing.

Perhaps  `original_hidden_dim` instead of `og_shared_hidden_states` and `transformed_hidden_dim` instead of `og_hidden_states`?

File: vllm/model_executor/layers/fused_moe/shared_fused_moe.py [hunk 4]
```diff
@@ -56,6 +58,25 @@ def gate(self) -> torch.nn.Module | None:
...
+
+        TODO: For latent MoE bandwidth optimization, fc2_latent_proj could be
+        moved inside SharedFusedMoE to all-reduce on the smaller latent
+        dimension.
+        """
+        if self._routed_input_transform is not None:
+            result = self._routed_input_transform(hidden_states)
+            # Handle modules that return tuples (e.g., ReplicatedLinear)
+            if isinstance(result, tuple):
+                return result[0]
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/shared_fused_moe.py:77]: This is to handle `ReplicatedLinear` returning (output, extra_bias), right? Could explain that explicitly in the comment?

**Review by tlrmchlsmth — COMMENTED** [review]
Could you add a unit test that exercises `apply_routed_input_transform`? I think something simple that compares the results `SharedFusedMoE` with the routed input transform happening inside vs outside the operation

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 2 above)
Comment (danielafrimi) [reply vllm/model_executor/layers/fused_moe/layer.py]: fixed using the context manager

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 1 above)
Comment (danielafrimi) [reply vllm/model_executor/layers/fused_moe/layer.py]: done

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 3 above)
Comment (danielafrimi) [reply vllm/model_executor/layers/fused_moe/layer.py]: done

File: vllm/model_executor/layers/fused_moe/shared_fused_moe.py (hunk 4 above)
Comment (danielafrimi) [reply vllm/model_executor/layers/fused_moe/shared_fused_moe.py:77]: yes sure

File: vllm/model_executor/layers/fused_moe/layer.py [hunk 5]
```diff
@@ -875,9 +900,17 @@ def _maybe_setup_shared_experts_stream(
         if use_shared_experts_stream:
             assert self.shared_experts_stream is not None
 
+            # For latent MoE: shared_experts need original [S, hidden_size],
+            # not the transformed [S, latent_size] used by routed experts.
+            shared_experts_input = (
+                self._shared_experts_input
+                if self._shared_experts_input is not None
+                else hidden_states
+            )
```
Comment (tlrmchlsmth) [inline vllm/model_executor/layers/fused_moe/layer.py]: nit: it'd be nice to factor this into a method since it appears twice

**Review by tlrmchlsmth — APPROVED** [review]

File: vllm/model_executor/layers/fused_moe/layer.py (hunk 5 above)
Comment (danielafrimi) [reply vllm/model_executor/layers/fused_moe/layer.py]: fixed it :) thanks
